### ISSUE TITLE
OOM in _setup_model_ckpt_dir: saving GT labels re-decodes every referenced video (4K → 58 GB peak, before data pipeline)

### ISSUE THREAD

**Issue by tom21100227** [talmolab/sleap-io#482]
## Summary

Training can OOM **during setup, before the data pipeline is ever built**, inside `ModelTrainer._setup_model_ckpt_dir()`. Saving the ground-truth `labels_gt.*.slp` copies re-opens and **decodes every video referenced by the labels** to recompute shapes that are already stored in the file's metadata. With many high-resolution source videos this balloons memory and gets the process killed.

In my case (finetune on a merged dataset whose `combined_train.slp` references **47 distinct 4K (3840×2160) videos**), this hit a **58 GB peak** and was OOM-killed. The last log line is always:

```
Setting up model ckpt dir: `.../finetune_final/centroid`...
```

i.e. it dies in `_setup_model_ckpt_dir`, *before* any dataset/dataloader is created — so this is **independent of `data_config.data_pipeline_fw`** (switching `torch_dataset` ↔ `torch_dataset_cache_img_disk` does not help).

## Environment

- sleap-nn `0.2.0` (also present on `main` — see line refs below)
- sleap-io `0.7.1`
- Python 3.13.13, PyTorch 2.12.1+cu130, CUDA 13.0, Linux
- Data: 800 train frames spread across 47 unique 4K grayscale `.mp4` videos (shapes already stored in `videos_json`, e.g. `[2517, 2160, 3840, 1]`); no embedded images.

## Root cause

1. Labels are loaded with `sio.load_slp(path)` (model_trainer.py:132/137/861) — `open_videos` defaults to `True`, so every `Video` gets an **open `MediaVideo` backend**.
2. `ModelTrainer.train()` calls `_setup_model_ckpt_dir()` (model_trainer.py:1128) **before** building the datasets (which happen at model_trainer.py:1130+).
3. `_setup_model_ckpt_dir()` (model_trainer.py:805) saves GT copies of train/val/test labels:
   ```python
   train_filtered.save(Path(ckpt_path) / f"labels_gt.train.{idx}.slp", restore_original_videos=False)  # :837
   val_filtered.save(  Path(ckpt_path) / f"labels_gt.val.{idx}.slp",   restore_original_videos=False)  # :841
   test_filtered.save( Path(ckpt_path) / f"labels_gt.test.{idx}.slp",  restore_original_videos=False)  # :868
   ```
4. In sleap-io, `Labels.save` → `write_videos` → `video_to_dict` takes the **`MediaVideo`** branch (slp.py:397) and reads `video.shape` / `video.grayscale` / `video.fps`:
   ```python
   elif type(video.backend) is MediaVideo:
       result["backend"] = {"type": "MediaVideo", "shape": video.shape, ...}   # slp.py:399-407
   ```
5. `Video._get_shape()` (model/video.py:200) decodes frames when the backend is open, and only uses the cached metadata shape **as an exception fallback**:
   ```python
   def _get_shape(self):
       try:
           return self.backend.shape          # backend OPEN -> img_shape() decodes frames
       except Exception:
           if "shape" in self.backend_metadata:
               return self.backend_metadata["shape"]   # correct shape, but only on error
           return None
   ```
   `VideoBackend.shape` → `img_shape` calls `read_test_frame()` + `detect_grayscale()` (video_reading.py), opening a `cv2.VideoCapture` on each 4K video. Because `keep_open=True` (default), **each opened 4K decoder context stays resident and is never freed** — so the cost accumulates across all referenced videos and all three GT saves (train + val + test).

The shape/grayscale/fps written to the GT file are already available in `video.backend_metadata` with **zero decoding** — the decode is pure waste.

## Evidence

**memray** (`memray run sleap-nn-train ...`), peak **58.1 GB** / 67 GB allocated. Two coexisting roots:
- `??? at ???` → **43.3 GB**, no Python stack = native ffmpeg/opencv decode buffers on decoder threads.
- Python branch, fully resolved:
  ```
  ModelTrainer.train (model_trainer.py)
   └ _setup_model_ckpt_dir
     └ Labels.save(embed=False) → write_videos → video_to_dict   ("shape": video.shape)
       └ Video.shape → _get_shape → VideoBackend.shape → _read_frame   ← decodes frames
  ```

**Minimal live measurement** — exact save path (`video_to_dict`) on 3 of the 47 videos:

| backends | per-video `video_to_dict` | shape correct? |
|---|---|---|
| **open** (`open_videos=True`, default / real run) | **+740 MB each, never released** (RSS 865 → 1603 → 2342 MB) | yes |
| **closed** (`open_videos=False`) | **+0.0 MB** (uses `backend_metadata["shape"]`) | yes — `[4986, 2160, 3840, 1]` |

47 videos × ~740 MB ≈ 35 GB for the train save alone; val + test saves open their own decoders on top → ~58 GB.

## Proposed fix

The GT label files are reference/eval artifacts — they only need video **metadata**, never decoded frames. So the save in `_setup_model_ckpt_dir` should not touch open backends.

**Recommended (sleap-nn):** save the GT copies with backends detached, so `video_to_dict` takes the `video.backend is None` path (`backend_metadata.copy()` → correct shape/grayscale/fps, zero decode). The filtered copies share `videos=labels.videos` with the live in-memory labels, so detach-then-restore (backends lazily reopen on later frame access):

```python
def _save_gt_metadata_only(labels, path):
    """Save GT labels using stored video metadata only (no frame decoding)."""
    backends = [v.backend for v in labels.videos]
    try:
        for v in labels.videos:
            v.backend = None        # -> video_to_dict writes backend_metadata.copy()
        labels.save(path, restore_original_videos=False)
    finally:
        for v, b in zip(labels.videos, backends):
            v.backend = b           # restore; data pipeline reopens lazily anyway
```

(Assumes `backend_metadata` is populated, which it is for labels loaded from `.slp`.) An equivalent alternative is to pre-seed `backend._cached_shape` from `backend_metadata["shape"]` so `VideoBackend.shape` short-circuits without decoding.

**Underlying (sleap-io)** — optional, worth a companion issue: `Video._get_shape()` / `video_to_dict()` should prefer `backend_metadata["shape"]` when present rather than forcing a decode via `backend.shape`, and metadata-only shape reads shouldn't leave `keep_open` 4K decoders resident.

Happy to open a PR with the sleap-nn fix (and a matching sleap-io one) if that's wanted.
